### PR TITLE
feat(logs): Add dbg() utility for generic debug logging (C++17) #205

### LINKED ISSUE #205: feat(logs): Add dbg() Utility for Generic Debug Logging (C++17)
#### **Summary**

Introduce a lightweight `dbg()` helper function to standardize debug logging across the graph engine. This utility converts various types (primitive, optional, pointer, etc.) into readable strings without requiring user-defined `toString()` or stream operators.

---

Current logging lacks consistent and meaningful context, especially for templated types like `VertexType` and `EdgeType`. Since user-defined types are not guaranteed to be printable, debugging becomes difficult.

A unified `dbg()` function will:

* Provide safe, generic string conversion
* Improve log readability
* Avoid compile-time constraints on user types
* Integrate cleanly with existing trait system

This function should work uniformly for all possible VertexType and EdgeType combinations, including primitives, strings, enums, pointers, optionals, Unweighted, and user-defined classes/structs. The goal is to ensure that every type used in the engine can be safely logged without requiring additional user effort or constraints.
#### **Proposed Implementation**
This has to be added in the ``Utils.hpp`` file. Try to work with printing edges and vertices IDs is the vertices and edges are complex.
This function should work uniformly for all possible VertexType and EdgeType combinations, including primitives, strings, enums, pointers, optionals, Unweighted, and user-defined classes/structs. The goal is to ensure that every type used in the engine can be safely logged without requiring additional user effort or constraints.
```cpp
template <typename T>
std::string dbg(const T& v) {
   
    // Primitive / string / enum
    if constexpr (is_primitive_enum_or_string_v<T>) {
      
    }

    // Unweighted edge
    else if constexpr (is_unweighted_v<T>) {
       
    }

    // Optional
    else if constexpr (is_optional_edge_v<T>) {
        
    }

    // Pointer
    else if constexpr (is_pointer_edge_v<T>) {
      
    }

    // Fallback (address-based identity)
    else {
        return "obj@" + std::to_string(reinterpret_cast<std::uintptr_t>(&v));
    }
}
```

---

#### **Usage Example**

```cpp
pHandler.log(LogLevel::WARN,
    "Edge not found: src(" + dbg(src) + "), dest(" + dbg(dest) + ")");
```

---

### Describe the Implementation Plan.

_No response_

### PR DESCRIPTION
**Pull request by Navyasree-ulava**
## Which issue does this PR close?

* Closes #205

---

## Rationale for this change

Debug logging in the graph engine is currently inconsistent for templated types like `VertexType` and `EdgeType`. Since user-defined types are not guaranteed to support `operator<<` or `toString()`, logging them directly can lead to compile-time errors or unreadable output.

This PR introduces a lightweight `dbg()` helper to safely convert any type into a readable string without imposing constraints on user-defined types.

---

### New

* `src/StorageEngine/DbgUtils.hpp`

  * Implements `CinderPeak::dbg<T>()` as a header-only inline template using C++17 `if constexpr`
  * Handles the following type categories:

    1. Primitives, `std::string`, enums → string conversion
    2. `Unweighted` → `"Unweighted"`
    3. `std::optional<T>` → recursive handling or `"nullopt"`
    4. Pointer types → recursive handling or `"null"`
    5. Types with `__id_` → `"id:<n>"`
    6. Fallback for unknown types → `"obj@<address>"`

### Modified

* `src/CinderPeak.hpp`

  * Added `#include "StorageEngine/DbgUtils.hpp"` to expose `dbg()` globally

### Tests

* `tests/unit/Utils/dbg_test.cpp`

  * Added 14 unit tests covering all branches and edge cases

---

## Design Notes

`dbg()` is implemented in a separate header (`DbgUtils.hpp`) instead of `Utils.hpp` to avoid include-order constraints between `Concepts.hpp` and `Utils.hpp`. This allows safe usage of project-defined type traits (`Traits::...`) without circular dependency issues.

---

## Are these changes tested?

Yes.

* Added 14 new unit tests
* All tests pass locally:

```
100% tests passed, 0 tests failed
```

* Additionally verified via manual usage with different type combinations

---

## Are there any user-facing changes?

Yes — `dbg()` is introduced as a new utility in `namespace CinderPeak`.

It is **fully backward compatible** and does not modify any existing APIs.

### Example usage

```cpp
pHandler.log(LogLevel::WARN,
    "Edge not found: src(" + dbg(src) + "), dest(" + dbg(dest) + ")");
```

### CODE REVIEW

File: src/StorageEngine/DbgUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * Does not use operator<<, streams, or toString().
+ * Never causes a compile-time error for unsupported types.
+ *
+ * @tparam T Any type used as VertexType or EdgeType.
+ * @param  v Value to convert.
+ * @return   A non-empty std::string suitable for log messages.
+ */
+template <typename T>
+inline std::string dbg(const T &v) {
```
Comment (SharonIV0x86) [inline src/StorageEngine/DbgUtils.hpp]: It would be better to move this function into the existing Utils.hpp file.
and add usage examples of the ``dbg`` function in [experimental examples](https://github.com/SharonIV0x86/CinderPeak/tree/main/examples/experimental)

Comment (Navyasree-ulava) [reply src/StorageEngine/DbgUtils.hpp]: Thanks for the suggestion!

I initially tried integrating `dbg()` directly into `Utils.hpp`, but it caused an include dependency issue since `Concepts.hpp` already includes `Utils.hpp`, while `dbg()` relies on the `Traits::...` predicates defined in `Concepts.hpp`.

To avoid that circular dependency cleanly, I moved the implementation into a dedicated `DbgUtils.hpp` header that can safely include both.

If you still prefer it strictly inside `Utils.hpp`, I can refactor the implementation to avoid the trait dependencies and move it there accordingly. Please let me know which approach you'd prefer.

I’ll also add experimental usage examples for `dbg()` in the next commit.

Comment (SharonIV0x86) [reply src/StorageEngine/DbgUtils.hpp]: Okay no issues, keep the dbg utility in a separate file.

**Comment by SharonIV0x86** [comment]
@Navyasree-ulava This looks fine. The CI is passing. 

It would be better to rename the file from ``DbgUtils.hpp`` to ``DebugUtils.hpp``. Once this is done we can work on getting it merged.

**Review by SharonIV0x86 — APPROVED** [review]
Approved.

Your PR has been approved and will be merged in sometime, till then if you like you can work on some other issue.

**Comment by SharonIV0x86** [comment]
@Navyasree-ulava Thank you for contributing, your PR has been merged. Kindly ⭐ the repository as it shows appreciation to the project maintainers.

**Comment by Navyasree-ulava** [comment]
Thank you! Really enjoyed contributing to the project and learning from the review process. I’ve starred the repository as well. Looking forward to contributing more.
